text/0250-Composite_Samplers.md

```diff
@@ -76,20 +76,20 @@ Each delegate sampler MUST be given a chance to participate in the sampling deci
 
 `Conjunction` is a composite sampler which takes two Samplers (delegates) as the arguments. These delegate samplers will be hereby referenced as First and Second.
 
-Upon invocation of its `shouldSample` method, the Conjunction sampler MUST invoke `shouldSample` method on the First sampler, passing the same arguments as received, and examine the received sampling Decision. Upon receiving `DROP` or `RECORD_ONLY` decision it MUST return the SamplingResult from the First sampler, and it MUST NOT proceed with querying the Second sampler. If the sampling decision from the First sampler is `RECORD_AND_SAMPLE`, the Conjunction sampler MUST invoke `shouldSample` method on the Second sampler.
+Upon invocation of its `shouldSample` method, the Conjunction sampler MUST invoke `shouldSample` method on the First sampler, passing the same arguments as received, and examine the received sampling Decision. Upon receiving `DROP` or `RECORD_ONLY` decision it MUST return the SamplingResult from the First sampler, and it MUST NOT proceed with querying the Second sampler. If the sampling decision from the First sampler is `RECORD_AND_SAMPLE`, the Conjunction sampler MUST invoke `shouldSample` method on the Second sampler, effectively passing the `TraceState` received from the First sampler as the parent trace state.
 If the sampling Decision from the Second sampler is `RECORD_AND_SAMPLE`, the Conjunction sampler MUST return a `SamplingResult` which is constructed as follows:
 
 - The sampling Decision is `RECORD_AND_SAMPLE`.
 - The set of span Attributes to be added to the `Span` is the sum of the sets of Attributes as provided by the First and the Second sampler.
 - The `TraceState` to be used with the new `Span` is obtained by cumulatively applying the potential modfications from the First and Second sampler, with special handling of the `th` sub-key (the sampling rejection `THRESHOLD`) for the `ot` entry as described below.
 
-If both First and Second samplers provided `th` entry in the returned `TraceState`, the resulting `TraceState` MUST contain `th` entry with the `THRESHOLD` being maximum of the `THRESHOLD`s provided by the First and Second samplers. Otherwise, the `th` entry MUST be removed.
+If both First and Second samplers provided `th` entry in the returned `TraceState`, and the value of the `THRESHOLD` from the First sampler is `0`, then the resulting `TraceState` MUST contain `th` entry with the `THRESHOLD`as provided by the Second sampler. Otherwise, the `th` entry MUST be removed.
```

Comment:
I do not understand how this definition allows to satisfy the last requirement in the given example without subsequent reservoir sampling. To my understanding, the sampling probability should be chosen first according to the rules
> - for root spans:
>   - drop all `/healthcheck` requests
>   - capture all `/checkout` requests
>   - capture 25% of all other requests
> - for non-root spans
>   - follow the parent sampling decision
>   - however, capture all calls to service `/foo` (even if the trace will be incomplete)

If the resulting sampling probabilities would lead to more than 1000 sampled spans/minute, the sampling probabilities should be adjusted (reduced) to satisfy the last rule
> * in any case, do not exceed 1000 spans/minute

In my interpretation this means that, if the rate of sampled spans would exceed 1000/min, the sampling probabilities as a result of the preceding rules need to be adjusted by the same reduction factor such that the limit is finally met. For that, the Second sampler must know the chosen threshold of the First sampler in order to be able to estimate the rate of sampled spans from its history (exponential smoothing) and also to apply the corresponding adjustment to the threshold. This is not possible, if the threshold is taken from the second sampler only if the threshold of the first sampler is zero.

Reply:
If we want to support the given example exclusively using consistent probability samplers, then the right solution is to use `EachOf` (as provided in the document) rather than `Conjunction`. In particular, using the rate limiting CP sampler with exponential smoothing as the Second delegate for `Conjunction` will not work correctly. The issue is not with the `th` value, but with overshooting the limit if the First delegate does not always decide to sample.
However, `Conjunction` is expected to work correctly if a reservoir (non CP) sampler is used there.

Admittedly, `Conjunction` looks strange. I'm trying to kill a few birds with a single stone here. It should be used with caution, and only when `EachOf` is not sufficient.

Reply:
> However, Conjunction is expected to work correctly if a reservoir (non CP) sampler is used there.

Unfortunately, reservoir sampling is not possible with the current sampler interface, as it requires an immediate sampling decision for every span. 

I am still wondering what is expected in the given example. For simplicity, assume that there are 2 categories `A` and `B` of spans. Assume there are 100 spans/min and 50% belong to `A` and 50% belong to `B`. Now we define to sample `A` with a probability of 60% and `B` with a probability of 40%. That means we would sample 30 spans/min from `A` and 20 spans/min from `B`.

If we do not want to sample more than 10 spans/min overall, what is our expectation?
1) Sample on average 5 spans/min from `A` and 5 spans/min from `B`?
2) Or, sample on average 6 spans/min from `A` and 4 spans/min from `B`?

With `EachOf`only the first can be achieved. For the second, we would need a rate-limiting consistent sampler, that adjusts the thresholds of the delegates by a factor that is determined from the history of thresholds using, for example, some kind of exponential smoothing.

Reply:
No, this is not the same. Just consider the case where the ratio between the number of A and B spans changes. For example, from the assumed 50:50 to 100:0. Then I still want to sample 10 spans/min without reconfiguring the samplers.

Reply:
> Unfortunately, reservoir sampling is not possible with the current sampler interface, as it requires an immediate sampling decision for every span.

An example of sampler that I had in mind here is Jaeger [RateLimitedSampler](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/RateLimitingSampler.java).

Reply:
> An example of sampler that I had in mind here is Jaeger [RateLimitedSampler](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/RateLimitingSampler.java).

I fear that this sampling approach cannot be turned into a consistent sampler, as the sampling probability is not known.

Reply:
> No, this is not the same. Just consider the case where the ratio between the number of A and B spans changes. For example, from the assumed 50:50 to 100:0. Then I still want to sample 10 spans/min without reconfiguring the samplers.

I've read your requirement as to provide the ratio of sampled spans 6:4. The solution works for a wide range of inputs, but obviously breaks if one of the span categories is critically underrepresented. A "fix" would require spilling some capacity from one category to another. How would your proposed adjustment of threshold handle this?

Reply:
A solution could be the following:

Dependent on the category, the first layer of samplers would choose the (normalized, decimal) thresholds `T_A=0.4` and `T_B=0.6` corresponding to 60% and 40% sampling probability, respectively. The chosen thresholds are further passed to the rate limiting sampler. If the ratio is 50:50, this sampler observes using exponential smoothing that the (weighted) average sampling probability is 50% and also that the overall rate of spans is 100/min. Therefore the sampler concludes, if the thresholds are kept as they are, that the rate of sampled spans would be 50/min. Therefore the rate limiting sampler decides to adjust the thresholds such that the goal of 10/min is achieved. It reduces the sampling probability by a factor 5, which means that the thresholds get adjusted to `T_A=0.88` and `T_B=0.92`, respectively.

If now the ratio changes suddenly, for example, to 100:0, sooner or later (dependent on the smoothing parameter) the sampler will realize that the average sampling probability has changed to 60% and therefore it has to reduce all sampling probabilities by a factor of 6 and to adjust all thresholds correspondingly.

Reply:
Got it. So the rate limiting sampler keeps two averages: incoming span rate and threshold value as reported by the delegate. And this gives _balancing_ behavior rather than _equalizing_.